source/common/stats/stats_impl.h

```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
+    flags_ |= Flags::Used;
+  }
+
+  void beginMerge() { current_active_ = 1 - current_active_; }
+
+  void merge() override {}
+
+  bool used() const override { return flags_ & Flags::Used; }
+
+  HistogramStatistics intervalStatistics() const override { return HistogramStatistics(); }
+
+  HistogramStatistics cumulativeStatistics() const override { return HistogramStatistics(); }
+
+  void merge(histogram_t* target) {
+    histogram_t* hist_array[1];
+    hist_array[0] = histograms_[1 - current_active_];
+    hist_accumulate(target, hist_array, ARRAY_SIZE(hist_array));
+    hist_clear(hist_array[0]);
+  }
+
+  Store& parent_;
+  int current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Log Linear Histogram implementation that is stored in the main thread.
+ */
+class HistogramParentImpl : public Histogram, public MetricImpl {
+public:
+  HistogramParentImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                      std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    interval_histogram_ = hist_alloc();
+    cumulative_histogram_ = hist_alloc();
+  }
+
+  ~HistogramParentImpl() {
+    hist_free(interval_histogram_);
+    hist_free(cumulative_histogram_);
+  }
+
+  // Stats::Histogram
+  void recordValue(uint64_t) override {}
+
+  bool used() const override {
+    bool any_tls_used = false;
+    for (auto tls_histogram : tls_histograms_) {
+      if (tls_histogram->used()) {
+        any_tls_used = true;
+        break;
+      }
+    }
+    return any_tls_used;
+  }
+
+  void addTlsHistogram(TlsHistogramSharedPtr hist_ptr) { tls_histograms_.emplace_back(hist_ptr); }
+
+  /**
+   * This method is called during the main stats flush process for each of the histogram. This
+   * method iterates through the Tls histograms and collects the histogram data of all of them
+   * in to "interval_histogram_". Then the collected "interval_histogram_" is merged to a
+   * "cumulative_histogram". More details about threading model at
+   * https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282.
+   */
+  void merge() override {
+    if (used()) {
+      hist_clear(interval_histogram_);
+      for (auto tls_histogram : tls_histograms_) {
+        tls_histogram->merge(interval_histogram_);
+      }
+      histogram_t* hist_array[1];
+      hist_array[0] = interval_histogram_;
+      hist_accumulate(cumulative_histogram_, hist_array, ARRAY_SIZE(hist_array));
+    }
+  }
+
+  HistogramStatistics intervalStatistics() const override {
+    return HistogramStatistics(interval_histogram_);
+  }
+
+  HistogramStatistics cumulativeStatistics() const override {
+    return HistogramStatistics(cumulative_histogram_);
+  }
+
+  Store& parent_;
+  std::list<TlsHistogramSharedPtr> tls_histograms_;
+  histogram_t* interval_histogram_;
```

Comment:
Yeah, I think in looking at this now I would make HistogramStatistics an interface, then where you need it, have an impl that wraps the circl implementation, which you can then return via const reference.